Send email digest for activities in Hypha for org level staff

theskumar opened the pull request:
Fixes #2882

This feature complements or provides an alternative to the slack notifications sent by the system to organization staff members.


- [x] 🏃🏼 management command to prepare and send the digest as an email `python manage.py send_staff_email_digest`
- [x] add email settings `ACTIVITY_DIGEST_RECIPIENTS_EMAILS`
- [x] Capture events of the last seven days only. 
- [x] Port all the staff messages sent to staff to the staff notification adaptor

![Screenshot 2022-10-06 at 11 06 46@2x](https://user-images.githubusercontent.com/236356/194298005-efb2c291-93e7-4317-bdcf-8e1e31437bb5.jpg)

![Screenshot 2022-10-06 at 11 08 47@2x](https://user-images.githubusercontent.com/236356/194298258-5d637fc8-c815-46e2-930e-9aac64bef9b5.jpg)


- add mdmail dependency
- add core mail send utils to wrap Django’s send_mail, which takes markdown or text as input.
- add new `ACTIVITY_DIGEST_RECIPIENT_EMAILS` setting

theskumar committed:
make activity -> messages options translatable

theskumar committed:
Refractor ADMIN user add or edit message to use slack adaptor

theskumar committed:
Implement management command
Use translations for email

theskumar commented:
@fourthletter please review the email format and copy

theskumar commented:
/rebased with main branch

frjo commented on requirements.txt:
```diff
@@ -37,6 +37,7 @@ drf-nested-routers==0.93.4
 drf-yasg==1.20.0
 environs==9.5.0
 gunicorn==20.1.0
+mdmail==0.1.3
```
This project does not seem to be active. Is that an issue? Could we take the bits we need and add them to Hypha?

theskumar replied:
I'll take a look.

theskumar replied:
Done

frjo commented on .env.sample:
```diff
@@ -0,0 +1 @@
+ACTIVITY_DIGEST_RECIPIENT_EMAILS=group1@example.com
```
I think `.env.example` is better, or vi rename `local.py.example`. Just so we are consistent.

theskumar replied:
I'll have it in the local.py.example, and if you like the .env.example approach, we can move it at once.

frjo commented:
When running the `send_staff_email_digest` command I get these errors:

~~~~
./hypha/venv/hypha/lib/python3.10/site-packages/django/db/models/fields/__init__.py:1358: RuntimeWarning: DateTimeField Event.when received a naive datetime (2022-09-30 00:00:00) while time zone support is active.
  warnings.warn("DateTimeField %s.%s received a naive datetime "
~~~~

Setting `USE_TZ = True` fixes it but we need timezone support.

frjo commented:
Then I get this error:

~~~~
Traceback (most recent call last):
  File "./hypha/manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "./hypha/hypha/apply/activity/management/commands/send_staff_email_digest.py", line 70, in handle
    send_mail(
  File "./hypha/hypha/core/mail.py", line 59, in send_mail
    mail.send()
  File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/mail/message.py", line 284, in send
    return self.get_connection(fail_silently).send_messages([self])
  File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/mail/backends/smtp.py", line 109, in send_messages
    sent = self._send(message)
  File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/mail/backends/smtp.py", line 125, in _send
    self.connection.sendmail(from_email, recipients, message.as_bytes(linesep='\r\n'))
  File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/smtplib.py", line 872, in sendmail
    self.ehlo_or_helo_if_needed()
  File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/smtplib.py", line 611, in ehlo_or_helo_if_needed
    if not (200 <= self.ehlo()[0] <= 299):
  File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/smtplib.py", line 451, in ehlo
    self.putcmd(self.ehlo_msg, name or self.local_hostname)
  File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/smtplib.py", line 378, in putcmd
    self.send(f'{s}{CRLF}')
  File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/smtplib.py", line 365, in send
    raise SMTPServerDisconnected('please run connect() first')
smtplib.SMTPServerDisconnected: please run connect() first
~~~~

The new `send_mail` function ignore what is set as `EMAIL_BACKEND` from what I understand?

Locally I use `EMAIL_BACKEND = 'django.core.mail.backends.filebased.EmailBackend'` (see `dev.py`) to have e-mails written as local files.

frjo commented on hypha/apply/activity/management/commands/send_staff_email_digest.py:
```diff
@@ -0,0 +1,80 @@
+def slack_message_to_markdown(msg):
+    def to_href(match_obj):
+        if match_obj.group() is not None:
+            return f"[{match_obj.group(2)}]({match_obj.group(1)})"
```
We prefer single quotes in all python code.

theskumar replied:
Sure. I'll update this.

frjo commented on hypha/apply/activity/adapters/slack.py:
```diff
@@ -65,6 +67,8 @@ class SlackAdapter(AdapterBase):
         MESSAGES.SUBMIT_REPORT: _('{user} has submitted a report for <{link}|{source.title}>'),
         MESSAGES.BATCH_DELETE_SUBMISSION: 'handle_batch_delete_submission',
+        MESSAGES.STAFF_ACCOUNT_CREATED: _('{user} has created a new account for <{link}|{source}>'),
+        MESSAGES.STAFF_ACCOUNT_EDITED: _('{user} has edited account for <{link}|{source}> that now has following roles: {roles}'),
```
In the long run I would prefer to make these generic, in Markdown perhaps? Then each staff message adapter can translate it to whatever is needed.

theskumar replied:
I agree, in terms of storage, I would prefer them to be JSON or JSON-serializable objects, compared to markdown.

frjo commented on hypha/core/mail.py:
```diff
@@ -0,0 +1,60 @@
+    mail.attach_alternative(content.html, "text/html")
+
+    mail.send()
+    return mail
```
This seems to ignore the `EMAIL_BACKEND` setting. Is it really needed just to send out HTML e-mails? Can we not extend the existing `send_mail()` with `EmailContent` instead?

theskumar replied:
`AnymailMessage` provides some standardized error codes per email provider, like mailgun, which is stored in the database, not here but when triggered via the celery task in the EmailAdapter. I tried to keep the interface the same. 

This function is essentially a re-write of the default `send_mail` provided by Django, which wrappers around the `EmailMessage` class to send text emails without HTML as attachment.

theskumar replied:
Done, implement as `MarkdownMail()` class with a send() function

frjo reviewed (changes requested):
A lot of nice functionality here! Only some concerns regarding sending out e-mails.

theskumar commented:
> When running the `send_staff_email_digest` command I get these errors:
> 
> ```
> ./hypha/venv/hypha/lib/python3.10/site-packages/django/db/models/fields/__init__.py:1358: RuntimeWarning: DateTimeField Event.when received a naive datetime (2022-09-30 00:00:00) while time zone support is active.
>   warnings.warn("DateTimeField %s.%s received a naive datetime "
> ```
> 
> Setting `USE_TZ = True` fixes it but we need timezone support.

I got these too. I believe that because the dates are stored in the DB without timezone info. IMO, timezone support should enabled and all the time datetime object should always contain a timezone info them.

theskumar commented:
> Then I get this error:
> 
> ```
> Traceback (most recent call last):
>   File "./hypha/manage.py", line 10, in <module>
>     execute_from_command_line(sys.argv)
>   File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
>     utility.execute()
>   File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/management/__init__.py", line 413, in execute
>     self.fetch_command(subcommand).run_from_argv(self.argv)
>   File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/management/base.py", line 354, in run_from_argv
>     self.execute(*args, **cmd_options)
>   File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/management/base.py", line 398, in execute
>     output = self.handle(*args, **options)
>   File "./hypha/hypha/apply/activity/management/commands/send_staff_email_digest.py", line 70, in handle
>     send_mail(
>   File "./hypha/hypha/core/mail.py", line 59, in send_mail
>     mail.send()
>   File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/mail/message.py", line 284, in send
>     return self.get_connection(fail_silently).send_messages([self])
>   File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/mail/backends/smtp.py", line 109, in send_messages
>     sent = self._send(message)
>   File "./hypha/venv/hypha/lib/python3.10/site-packages/django/core/mail/backends/smtp.py", line 125, in _send
>     self.connection.sendmail(from_email, recipients, message.as_bytes(linesep='\r\n'))
>   File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/smtplib.py", line 872, in sendmail
>     self.ehlo_or_helo_if_needed()
>   File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/smtplib.py", line 611, in ehlo_or_helo_if_needed
>     if not (200 <= self.ehlo()[0] <= 299):
>   File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/smtplib.py", line 451, in ehlo
>     self.putcmd(self.ehlo_msg, name or self.local_hostname)
>   File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/smtplib.py", line 378, in putcmd
>     self.send(f'{s}{CRLF}')
>   File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/smtplib.py", line 365, in send
>     raise SMTPServerDisconnected('please run connect() first')
> smtplib.SMTPServerDisconnected: please run connect() first
> ```
> 
> The new `send_mail` function ignore what is set as `EMAIL_BACKEND` from what I understand?
> 
> Locally I use `EMAIL_BACKEND = 'django.core.mail.backends.filebased.EmailBackend'` (see `dev.py`) to have e-mails written as local files.

Ok. I used an STMP email server via .env configs and didn't change anything else in the `local.py`. Can you share your local.py config for me to test it out?

frjo commented:
In local.py:
~~~~
# Write e-mails to local files.
LOCAL_FILE_EMAIL = True
~~~~

That activate this in dev.py:

~~~~
# E-mail to local files.
if LOCAL_FILE_EMAIL:
    EMAIL_BACKEND = 'django.core.mail.backends.filebased.EmailBackend'
    EMAIL_FILE_PATH = BASE_DIR + '/var/mail'
~~~~

theskumar commented:
For `EMAIL_BACKEND` seems to be corrupted somehow. Can you open your shell and tell me the value of your EMAIL_BACKEND?

```
$ python manage.py shell

>> from django.conf import settings
>> settings.EMAIL.BACKEND
```

If the `DJANGO_SETTINGS_MODULE` is set to `hypha.settings.production`, which is the default in the `manage.py` the `LOCAL_FILE_EMAIL` is not honored.

theskumar commented:
Also, please check if you set the value of `DJANGO_SETTINGS_MODULE` to `"hypha.settings.dev"`

It defaults to `"hypha.settings.production"`.

sandeepsajan0 reviewed (approved):
Works great for me. 

<img width="1234" alt="image" src="https://user-images.githubusercontent.com/23638629/194869625-b614562e-d002-4445-b180-cbced3feab98.png">

frjo commented on hypha/apply/activity/management/commands/send_staff_email_digest.py:
```diff
@@ -0,0 +1,88 @@
+        if total_count:
+            email = MarkdownMail('emails/activity_summary.md')
+            email.send(
+                to=settings.ACTIVITY_DIGEST_RECIPIENT_EMAILS,
```
Is ACTIVITY_DIGEST_RECIPIENT_EMAILS a list or a comma separated string?

theskumar replied:
`ACTIVITY_DIGEST_RECIPIENT_EMAILS` comes via `env.list()`  which converts a comma separated string to python list. 

`ACTIVITY_DIGEST_RECIPIENT_EMAILS` is a list at the end.

frjo commented:
It works for me now, see https://github.com/HyphaApp/hypha/issues/3004.

frjo commented on requirements.txt:
```diff
@@ -37,6 +37,7 @@ drf-nested-routers==0.93.4
 drf-yasg==1.20.0
 environs==9.5.0
 gunicorn==20.1.0
+Markdown==3.4.1
```
Can we make use of the `mistune` packages that we already have installed instead of adding another Markdown package?

I picked `mistune` for its performance and it seems to work well.

theskumar replied:
Done in https://github.com/HyphaApp/hypha/pull/2991/commits/2b4ea1be0d625f1cf6858a3237ab7a5c2d8711d1

frjo commented on hypha/apply/activity/templates/emails/activity_summary.md:
```diff
@@ -0,0 +1,29 @@
+
+{% if submissions %}**{% trans "Submissions" %}**
+{% for msg in submissions %}
+- {{ msg.content_markdown }} • {{ msg.event.when }}{% endfor %}{% endif %}
```
After each "•" there is a nonbreaking space character (0xa0). I think that should be a normal space.

theskumar replied:
Fixed in https://github.com/HyphaApp/hypha/pull/2991/commits/68956bda895341f8c5ea81c13d0937102efbd04b

frjo commented on hypha/apply/activity/templates/emails/activity_summary.md:
```diff
@@ -0,0 +1,29 @@
+
+{% trans "Kind Regards" %},  
+{% trans "Team" %}  
+[{{org.NAME}}]{% if org.URL %}({{org.URL}}){% endif %}
```
Would it make sense here to reuse the "base.html" template? Any maybe the template could reside with the other e-mail templates in "messages/email" dir?

theskumar replied:
Done

frjo reviewed (changes requested):
This is looking really nice. Some minor issues I commented on separately.

theskumar committed:
Use mistune for markdown render, create markdown_to_html util

theskumar committed:
Fix whitespace character

theskumar committed:
Use base email template

theskumar committed:
Add space character in footer separator

frjo commented on hypha/apply/activity/templates/messages/email/activity_summary.md:
```diff
@@ -0,0 +1,28 @@
+{% extends "messages/email/base.html" %}
+{% load humanize i18n %}
+
+{% block salutation %}{% trans "Dear" %},{% endblock %}
```
Wondering if some kind of header makes more sense here, we do not have a specific user to say "Dear nn" too.

What do you think about setting it too `# {% trans "Activities Summary" %}` or similar?

theskumar replied:
That would work; let me know if this sounds better.

---

Activities Summary
--------------------

There are two new activities:


```
Activities Summary
--------------------

There is a new activity:

```

theskumar replied:
![Screenshot 2022-10-14 at 06 12 24@2x](https://user-images.githubusercontent.com/236356/195775499-804a0343-e8c0-489f-8ec7-e095343f89a4.jpg)

theskumar replied:
Updated in https://github.com/HyphaApp/hypha/pull/2991/commits/67cb1a9e80433fe7ba795771846cee34bea7e293

frjo replied:
Good improvement!

theskumar committed:
Update email template copy

theskumar committed:
Add ‘Auto-submitted: auto-generated’ header to email.

https://www.iana.org/assignments/auto-submitted-keywords/auto-submitted-keywords.xhtml

theskumar commented:
@frjo I've also added 'Auto-Submitted' header with 'auto-generated' value

> Indicates that a message was generated by an automatic process, and is not a direct response to another message.

https://github.com/HyphaApp/hypha/pull/2991/commits/8a13212dfaa79ee72b494587a97be13593ecc3c5

https://www.iana.org/assignments/auto-submitted-keywords/auto-submitted-keywords.xhtml

theskumar commented:
/rebased
